helper/fairshare/jobmanager.go

```diff
@@ -179,10 +182,81 @@ func (j *JobManager) getNextJob() Job {
 	}
 
 	if j.queues[queueID].Len() == 0 {
+		// we remove the empty queue, but we don't remove the worker count
+		// in case we are still working on previous jobs from this queue.
+		// worker count cleanup is handled in j.decrementWorkerCount
 		j.removeLastQueueAccessed()
 	}
 
-	return out.(Job)
+	return jobRaw.(Job), queueID
+}
+
+// returns the next queue to assign work from, and a bool if there is a queue
+// that can have a worker assigned. if there is work to be assigned,
+// j.lastQueueAccessed will be updated to that queue.
+// note: this must be called with j.l held
+func (j *JobManager) getNextQueue() (string, bool) {
+	var nextQueue string
+	var canAssignWorker bool
+
+	// ensure we loop through all existing queues until we find an eligible
+	// queue, if one exists.
+	queueIdx := j.nextQueueIndex(j.lastQueueAccessed)
+	for i := 0; i < len(j.queuesIndex); i++ {
+		potentialQueueID := j.queuesIndex[queueIdx]
+
+		if !j.queueWorkersSaturated(potentialQueueID) {
+			nextQueue = potentialQueueID
+			canAssignWorker = true
+			j.lastQueueAccessed = queueIdx
+			break
+		}
+
+		queueIdx = j.nextQueueIndex(queueIdx)
+	}
+
+	return nextQueue, canAssignWorker
+}
+
+// get the index of the next queue in round-robin order
+// note: this must be called with j.l held
+func (j *JobManager) nextQueueIndex(currentIdx int) int {
+	return (currentIdx + 1) % len(j.queuesIndex)
+}
+
+// returns true if there are already too many workers on this queue
+// note: this must be called with j.l held (at least for read).
+// note: we may want to eventually factor in queue length relative to num queues
+func (j *JobManager) queueWorkersSaturated(queueID string) bool {
+	numActiveQueues := float64(len(j.queues))
+	numTotalWorkers := float64(j.workerPool.numWorkers)
+	maxWorkersPerQueue := math.Ceil(0.9 * numTotalWorkers / numActiveQueues)
```

Comment:
it might make sense to reduce the 0.9, or ration workers out based on number of available workers (rather than total workers)?

Reply:
could you explain why this suggestion? i think that would do an integer divide before a float multiplication (and order of operations makes the math correct as is)?

see https://play.golang.org/p/48DlknViyjk for an example

Reply:
i'm going to mark this as resolved without changing it, but happy to discuss if there's something i've missed!